File: src/common/weekly/415/3289.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
var getSneakyNumbers = function (nums) {
  let xor = 0;
  for (let i = 0; i < nums.length; i++) {
    xor ^= i ^ nums[i];
  }
  let n = nums.length - 2;
  xor ^= n;
  xor ^= n + 1;
  console.log(xor);
  // 现在oxr里面包含的是这俩重复的数字
  let lowBit = xor & -xor;
  console.log(lowBit);
  let res = [0, 0];
  for (let i = 0; i < nums.length; i++) {
    let x = nums[i];
    if (i < n) {
      ans[(i >> shift) & 1] ^= i;
    }
    res[lowBit & x ? 1 : 0] ^= x;
  }
  return res;
};

File: src/common/weekly/415/3289.py
class Solution:
    def getSneakyNumbers(self, nums: List[int]) -> List[int]:
        n = len(nums) - 2
        xor_all = n ^ (n + 1)  # n 和 n+1 多异或了
        for i, x in enumerate(nums):
            xor_all ^= i ^ x
        shift = xor_all.bit_length() - 1

        ans = [0, 0]
        for i, x in enumerate(nums):
            if i < n:
                ans[i >> shift & 1] ^= i
            ans[x >> shift & 1] ^= x
        return ans



class Solution:
    def getSneakyNumbers(self, nums: List[int]) -> List[int]:
        n = len(nums) - 2
        a = -n * (n - 1) // 2
        b = -n * (n - 1) * (n * 2 - 1) // 6
        for x in nums:
            a += x
            b += x * x
        x = int((a - sqrt(b * 2 - a * a)) / 2)
        return [x, a - x]
